### DISCUSSION TITLE
Authentication using cookies and Next.js Server Actions

### DISCUSSION THREAD

**Discussion by heloineto** [pocketbase/pocketbase#3201] [Q&A] [answered]
Hi there. I'm trying to set up authentication on my Next.js 13 app using PocketBase and Server Actions

Here's what I got so far:

```ts
'use server';

import Pocketbase from 'pocketbase';
import { cookies } from 'next/headers';

type LoginFormValues = {
  email: string;
  password: string;
}

export default async function login(values: LoginFormValues) {
  // I believe it's a bad idea to instantiate the client on every run, will fix once I get this to work
  const pocketbase = new Pocketbase('http://127.0.0.1:8090');

  await pocketbase.collection('users').authWithPassword(values.email, values.password);

  const authCookie = pocketbase.authStore.exportToCookie({
    httpOnly: false,
  });

  /**
   * I believe this is wrong, because the `exportToCookie` function returns something like this:
   * "pb_auth=<URI_ENCODED_COOKIE_VALUE>; Path=/; Expires=Thu, 07 Sep 2023 12:55:53 GMT; Secure; SameSite=Strict"
   * 
   * From what I read on the docs, I should only pass the <URI_ENCODED_COOKIE_VALUE> value as the second argument of the
   * `cookies().set` call.
   */
  cookies().set('pb_auth', authCookie);
}
```

Is there any way to get only the <URI_ENCODED_COOKIE_VALUE>, so I can pass it as the second argument of `cookies().set`?
I would rather not use Regex or other hacky solution and would love the cookie to have the same optional attributes as well, is there a better way to do this?

**Comment by ganigeorgiev** [comment, marked as answer]
What you refer by _"URI_ENCODED_COOKIE_VALUE"_ is basically:
```js
// (the model is optional)
JSON.stringify({token: pb.authStore.token, model: pb.authStore.model})
```

If you really need only the cookie value+attributes without the name, you can do something like:
```js
pb.authStore.exportToCookie().substring('pb_auth='.length)
```

But keep in mind that I'm not sure if Next.js allow the cookie attributes to be passed together as part of the value.

`exportToCookie` (and `loadFromCookie`) are just helpers that works with plain cookies. You don't have to use it if you are going to rely on the Next.js cookies abstraction.

**Reply by heloineto** [reply]
Thank you so much for your answer!

I'll study the js-sdk code to better understand how this all works, if I can figure this out I'll create a demo repo.

**Reply by tigawanna** [reply]
I set something up a while back before the new SDK but most of it will still work
 https://dev.to/tigawanna/nextjs13-app-directory-authentication-with-pocketbase-2p6d

**Reply by heloineto** [reply]
Hi! I read your article and found it very insightful. However, I believe my implementation differs a bit from yours since I'm trying to keep all the authentication logic server-side, without using the SDK on the client.

**Reply by tigawanna** [reply]
Yeah , I wanted to go that route too but the cookie and cache invalidation was janky and vaguely documented back then especially when using server actions . 
The only thing you need t look out for is the output from cookie.get it outputs a key value object that need to be serialized into a cookie string

**Reply by heloineto** [reply]
I created an example repo for the setup I went with: https://github.com/heloineto/nextjs-13-pocketbase-auth-example

I hope it helps future folks. Let me know if there's anything I can do better!
